src/edu/stanford/nlp/sempre/DerivationGroup.java

```diff
@@ -0,0 +1,32 @@
+package edu.stanford.nlp.sempre;
+
+import java.util.List;
+
+/**
+ * Represents one or two lists of derivations.
+ *
+ * The motivation is to group derivations based on type compatibility.
+ * For example, when building (and __ __), considering all pairs of derivations
+ * is time-wasting since a lot of pairs don't type-check. We instead group
+ * derivations by type, and only apply the rule to the pairs that type-check.
+ *
+ * During parsing, for each DerivationGroup:
```

Comment:
Is the difference between the two cases just rules that take one argument versus two?